File: learning-core/executable/idempotent-message/cpp/test_idempotent_message.cpp
#include <cassert>

#include "idempotent_message.cpp"

int main() {
    ConsumerState state;
    assert(handle(state, "m1", 10));
    assert(!handle(state, "m1", 10));
    assert(state.total == 10);

    assert(handle(state, "m2", 5));
    assert(state.total == 15);
}

File: learning-core/executable/idempotent-message/python/test_idempotent_message.py
from idempotent_message import ConsumerState, handle


def test_duplicate_message_does_not_repeat_side_effect() -> None:
    state = ConsumerState()
    assert handle(state, "m1", 10) is True
    assert handle(state, "m1", 10) is False
    assert state.total == 10


def test_distinct_messages_are_applied() -> None:
    state = ConsumerState()
    handle(state, "m1", 10)
    handle(state, "m2", 5)
    assert state.total == 15
